File: DataStructures/BloomFilter.py

class BloomFilter(object):

    def __init__(self, size=10):

        self.store = [False] * size

    def get_hashes(self, item):

        hashes = [
            hash(item),
            hash(item + " "),
            hash(item + "!")
        ]

        return hashes

    def add(self, item):

        for hash_ in self.get_hashes(item):

            index = hash_ % len(self.store)

            self.store[index] = True

    def might_contain(self, item):

        for hash_ in self.get_hashes(item):

            index = hash_ % len(self.store)

            if not self.store[index]:
                return False

        return True

def main():

    for bloom_class in [BloomFilter]:

        bloom = bloom_class()

        bloom.add("A")
        bloom.add("B")
        bloom.add("C")

        print(bloom.might_contain("A"))
        print(bloom.might_contain("Z"))

        print("------")

if __name__ == '__main__':
    main()

File: DataStructures/BloomFilter.js
"use strict";

String.prototype.hashCode = function() { // https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript-jquery
    var hash = 0,
        i, chr;
    if (this.length === 0) return hash;
    for (i = 0; i < this.length; i++) {
        chr = this.charCodeAt(i);
        hash = ((hash << 5) - hash) + chr;
        hash |= 0; // Convert to 32bit integer
    }
    return hash;
};

class BloomFilter {

    constructor(size = 10) {

        this.store = [];
        while (this.store.push(false) < size);

    }

    getHashes(item) {

        let hashes = [
            item.hashCode(),
            (item + "").hashCode(),
            (item + "!").hashCode()
        ];

        return hashes;

    }

    add(item) {

        for (let hash of this.getHashes(item)) {

            let index = hash % this.store.length;

            this.store[index] = true;

        }

    }

    mightContain(item) {

        for (let hash of this.getHashes(item)) {

            let index = hash % this.store.length;

            if (!this.store[index]) {
                return false;
            }

        }

        return true;

    }

}

for (let bloomClass of [BloomFilter]) {

    let bloom = new bloomClass();

    bloom.add("A");
    bloom.add("B");
    bloom.add("C");

    console.log(bloom.mightContain("A"));
    console.log(bloom.mightContain("Z"));

    console.log("------");

}
